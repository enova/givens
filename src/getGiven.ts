import getGivenFunc from './getGivenFunc';
import { given } from './types';

export default function getGiven<T>(): given<T> {
  const givenFunc = getGivenFunc();

  return Object.defineProperties(givenFunc, {
    __props__: { value: {} },
    __cache__: { value: {} },
    __trace__: { value: [] },
  }) as given<T>;
}
