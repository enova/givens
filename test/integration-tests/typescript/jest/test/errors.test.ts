import assert from 'assert';
import getGiven from 'givens';

const given = getGiven();

describe('illegal prop', () => {
  assert.throws(
    () => given('__props__', () => 'value'),
    (err) => err.message === 'givens: key "__props__" is not allowed',
  );

  it('does stuff', () => {
    expect(true);
  });
});

describe('illegal call location', () => {
  describe('in beforeEach method', () => {
    beforeEach(() => {
      assert.throws(
        () => given('test', () => 'foo'),
        (err) => err.message === 'givens: cannot call given from a lifecycle hook',
      );
    });

    it('breaks', () => undefined);
  });

  describe('in afterEach method', () => {
    afterEach(() => {
      assert.throws(
        () => given('test', () => 'foo'),
        (err) => err.message === 'givens: cannot call given from a lifecycle hook',
      );
    });

    it('breaks', () => undefined);
  });

  describe('in beforeAll method', () => {
    beforeAll(() => {
      assert.throws(
        () => given('test', () => 'foo'),
        (err) => err.message === 'givens: cannot call given from a lifecycle hook',
      );
    });

    it('breaks', () => undefined);
  });

  describe('in afterAll method', () => {
    afterAll(() => {
      assert.throws(
        () => given('test', () => 'foo'),
        (err) => err.message === 'givens: cannot call given from a lifecycle hook',
      );
    });

    it('breaks', () => undefined);
  });

  describe('in test', () => {
    it('breaks', () => {
      expect(() => given('test', () => 'foo')).toThrowErrorMatchingSnapshot();
    });
  });
});

describe('with recursive given variables', () => {
  given('recursive1', () => {
    const string1 = 'recursive1 => ';
    const string2 = given.recursive2;
    return string1 + string2;
  });
  given('recursive2', () => {
    const string1 = 'recursive2 => ';
    const string2 = given.recursive1;
    return string1 + string2;
  });
  given('nonRecursive', () => given.recursive1);

  it('blows up', () => {
    expect(() => given.nonRecursive).toThrowErrorMatchingSnapshot();
  });
});
