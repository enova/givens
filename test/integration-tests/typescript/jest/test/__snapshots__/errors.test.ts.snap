// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`illegal call location in test breaks 1`] = `"givens: cannot call given from a test"`;

exports[`with recursive given variables blows up 1`] = `
"givens: recursive variable recursive1 detected
trace: recursive1 => recursive2 => recursive1"
`;
